apps/internal/base/base.go

```diff
@@ -0,0 +1,299 @@
+	grantedScopes := strings.Split(storageTokenResponse.AccessToken.Scopes, scopeSeparator)
+
+	// Checking if there was an ID token in the cache; this will throw an error in the case of confidential client applications.
+	var idToken accesstokens.IDToken
+	if !storageTokenResponse.IDToken.IsZero() {
+		idToken, err = accesstokens.NewIDToken(storageTokenResponse.IDToken.Secret)
+		if err != nil {
+			return AuthenticationResult{}, err
+		}
+	}
+	return AuthenticationResult{account, idToken, accessToken, expiresOn, grantedScopes, nil}, nil
+}
+
+// CreateAuthenticationResult creates an AuthenticationResult.
+// TODO(jdoak): (maybe, we did a refactor): Make this a method on TokenResponse() that takes only 1 arge, Account.
+func CreateAuthenticationResult(tokenResponse accesstokens.TokenResponse, account shared.Account) (AuthenticationResult, error) {
+	if len(tokenResponse.DeclinedScopes) > 0 {
+		return AuthenticationResult{}, fmt.Errorf("token response failed because declined scopes are present: %s", strings.Join(tokenResponse.DeclinedScopes, ","))
+	}
+	return AuthenticationResult{
+		Account:       account,
+		IDToken:       tokenResponse.IDToken,
+		AccessToken:   tokenResponse.AccessToken,
+		ExpiresOn:     tokenResponse.ExpiresOn,
+		GrantedScopes: tokenResponse.GrantedScopes,
+	}, nil
+}
+
+// Client is a base client that provides access to common methods and primatives that
+// can be used by multiple clients.
+type Client struct {
+	Token   *requests.Token
+	manager manager // *storage.Manager or fakeManager in tests
+
+	AuthParams    authority.AuthParams // DO NOT EVER MAKE THIS A POINTER! See "Note" in New().
+	cacheAccessor cache.ExportReplace
+}
+
+// New is the constructor for Base.
+func New(clientID string, authorityURI string, cacheAccessor cache.ExportReplace, token *requests.Token) (Client, error) {
+	authInfo, err := authority.NewInfoFromAuthorityURI(authorityURI, true)
+	if err != nil {
+		return Client{}, err
+	}
+	authParams := authority.NewAuthParams(clientID, authInfo)
+
+	return Client{ // Note: Hey, don't even THINK about making Base into *Base. See "design notes" in public.go and confidential.go
+		Token:         token,
+		AuthParams:    authParams,
+		cacheAccessor: noopCacheAccessor{},
+		manager:       storage.New(token),
+	}, nil
+}
+
+// AuthCodeURL creates a URL used to acquire an authorization code.
+func (b Client) AuthCodeURL(ctx context.Context, clientID, redirectURI string, scopes []string, authParams authority.AuthParams) (string, error) {
+	endpoints, err := b.Token.ResolveEndpoints(ctx, authParams.AuthorityInfo, "")
+	if err != nil {
+		return "", err
+	}
+
+	baseURL, err := url.Parse(endpoints.AuthorizationEndpoint)
+	if err != nil {
+		return "", err
+	}
+
+	v := url.Values{}
+	v.Add("client_id", clientID)
+	v.Add("response_type", "code")
+	v.Add("redirect_uri", redirectURI)
+	v.Add("scope", strings.Join(scopes, " "))
+
+	// There were left over from an implementation that didn't use any of these.  We may
```

Comment:
some of these are needed for proper implementation of the auth code flow, some are needed for parameters that a developer would pass in for special behaviors. we should have work items to address those.

Reply:
@abhidnya13 can you confirm an entry to track this?

Reply:
Tracking this here: https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/121

Reply:
Being tracked by abhidnya13